### PR TITLE
Add OCR and Circle detection

### PR DESCRIPTION
**Pull request by steveoh**
replaces #5

### CODE REVIEW

**Comment by steveoh** [comment]
@eneemann i think this is in a good spot now. We'll just need to fool with the ocr part before we merge. Take a look at the changes and let me know what you think.

File: row.py [hunk 1]
```diff
@@ -81,9 +100,215 @@ def convert_pdf_to_pil(pdf_as_bytes):
...
+            circles (array): Circle locations returned from cv2.HoughCircles algorithm
+            out_dir (Path): The output directory for cropped images of detected circles
+            file_path (Path): The location of the image file
+            cv2_image (numpy.ndarray): The image as a numpy array
+            height (number): The height of original image
+            width (number): The width of original image
+            inset_distance (number): The inset distance in pixels to aid image cropping
+
+    Returns:
+        list: a list of images and the count of images
```
Comment (eneemann) [inline row.py:270]: Isn't this just returning a list of the images and not the count?  Though the count can be determined by the length of the list.

File: row.py [hunk 2]
```diff
@@ -81,9 +100,215 @@ def convert_pdf_to_pil(pdf_as_bytes):
...
+    if not item_path.name.casefold().endswith(("jpg", "jpeg", "tif", "tiff", "png")):
+
+        return []
+
+    logging.debug("detecting circles in %s", item_path)
+    #: read in image from bytes
+    byte_img = item_path.read_bytes()
+    img = cv2.imdecode(np.frombuffer(byte_img, dtype=np.uint8), 1)  # 1 means flags=cv2.IMREAD_COLOR
+
+    if output_path:
```
Comment (eneemann) [inline row.py:130]: We can probably get rid of L130-134.  This makes a copy of the original image in the output folder, which I mainly did to make sure it looked correct when I read it in using the read_bytes/cv2.imdecode/np.frombuffer method.

**Review by eneemann — COMMENTED** [review]
Overall, I think this looks pretty good.  We did miss a few late commits that I had in my original PR that would probably be worth including in this one.
[fix: prevent crop failures on edges](https://github.com/agrc/udot-parcel-ml/pull/5/commits/3312a01aff425c38c08e33f63bc778dc0992547a)
[feat: process images to improve OCR](https://github.com/agrc/udot-parcel-ml/pull/5/commits/f9cc855b31863791f45b039c4b7b8da35f4130b8)

File: row.py (hunk 1 above)
Comment (eneemann) [inline row.py:270]: Isn't this just returning a list of the images and not the count? Though the count can be determined by the length of the list.

Comment (steveoh) [reply row.py:270]: yeah not sure where the count came from... we can reove that

File: row.py [hunk 3]
```diff
@@ -81,9 +100,215 @@ def convert_pdf_to_pil(pdf_as_bytes):
...
+        outfile = str(output_path / "test_1.jpg")  #: checking to make sure the image looks right
+
+        logging.debug(outfile)
+        cv2.imwrite(outfile, img)
```
[the remark this answers is not in the document]
Comment (steveoh) [reply row.py:134]: ok yeah we should remove this

File: row.py [hunk 4]
```diff
@@ -209,49 +211,45 @@ def get_circles(item_path, output_path):
...
-        ocr_text = pytesseract.image_to_string(ocr_gray, config=OCR_CONFIG)
-        ocr_text = ocr_text.replace("\n", "").replace("\t", "").replace(" ", "")
-        df_new = pd.DataFrame(
-            {
-                "Number": [int(ocr_count)],
-                "Filename": [ocr_file],
-                "Parcel": [ocr_text],
-            }
-        )
+    image_channels = cv2.imdecode(np.frombuffer(image, dtype=np.uint8), 1)  # 1 means flags=cv2.IMREAD_COLOR
```
Comment (eneemann) [inline row.py:219]: Are we calling np.frombuffer twice?  Once in row_cli.py before get_characters() is called (L92), and then again here inside the get_characters() function?  If we call it outside the function, I don't think we need it inside the function, too.

File: .github/workflows/pull_request.yml [hunk 5]
```diff
@@ -22,8 +22,7 @@ jobs:
           python-version: 3.9
 
       - name: 🖇️ Install dependencies
-        run: sudo apt-get update && sudo apt-get install libpoppler-dev poppler-utils -y
-
+        run: sudo apt-get update && sudo apt-get install libpoppler-dev poppler-utils tesseract-ocr libtesseract-dev -y
```
Comment (eneemann) [inline .github/workflows/pull_request.yml:25]: I'm not sure if `libtesseract-dev` is necessary, but it's probably not a problem

File: README.md [hunk 6]
```diff
@@ -26,3 +25,15 @@ Get files from bucket
...
+Detect circles in an image
+
+```py
+python row_cli.py detect circles "path_to_image.jpg"
+```
+
+OCR all image in a directory
+
+```py
+python row_cli.py detect characters --ocr-directory="directory_path"
```
Comment (eneemann) [inline README.md:38]:
```suggestion
python row_cli.py detect characters --ocr-directory="directory_path"
python row_cli.py detect characters ./test-data/crop_808_A.jpg
```

row_cli.py detect characters <file_name> [--ocr-directory=directory]

Comment (eneemann) [reply README.md:38]: Include an example on a single image

File: row_test.py [hunk 7]
```diff
@@ -103,3 +102,59 @@ def test_get_job_files_returns_the_page_size_for_the_second_page(storage_client)
...
+        file_name = item_path.name
+
+        image_array = np.frombuffer(item_path.read_bytes(), dtype=np.uint8)
+
+        characters = row.get_characters(image_array)
+
+        file_name = file_name.replace("crop_", "").replace("_", ":").replace(".jpg", "")
+
+        assert characters is not None
+        assert characters == file_name
```
Comment (eneemann) [inline row_test.py:160]:
```suggestion
        expected_characters = file_name.replace("crop_", "").replace("_", ":").replace(".jpg", "")

        assert characters is not None
        assert characters == expected_characters
```
It's a little confusing to reuse the `file_name` variable, I recommend changing this to `expected_characters`

**Comment by eneemann** [comment]
General comment on flow of data, I think I'm a little confused because of how we've mostly been focused on the CLI. If `row.get_circles` returns a list of images, then in `row_run.py` we'll need to iterate through the list and call `row.get_characters` on each individual image with circles.  But in the CLI, we don't really handle the list of returned images from `get_circles`, we only output them to a directory if one is provided. Is that right?

**Review by eneemann — CHANGES_REQUESTED** [review]
I have a few comments and minor changes to review before merging, but I think we're getting close on this one.

File: row.py (hunk 4 above)
Comment (steveoh) [reply row.py:219]: good catch. I didn't like having numpy in the cli so let's keep it all in row and send in the bytes only.

File: .github/workflows/pull_request.yml (hunk 5 above)
Comment (steveoh) [reply .github/workflows/pull_request.yml:25]: i copied this from the docs. we can try removing it if you'd like to see if the build/tests still pass. I'm not too worried about it.

**Comment by steveoh** [comment]
> General comment on flow of data, I think I'm a little confused because of how we've mostly been focused on the CLI. If `row.get_circles` returns a list of images, then in `row_run.py` we'll need to iterate through the list and call `row.get_characters` on each individual image with circles. But in the CLI, we don't really handle the list of returned images from `get_circles`, we only output them to a directory if one is provided. Is that right?

yes that is correct. The CLI is there to test the functionality of the row module. Think of it as a development tool. It's not built to solve the entire problem space we are tasked with. That will be done in the `row_run` like you mentioned.

**Review by eneemann — APPROVED** [review]
I think it looks good!

**Pull request merged**